prov/tcp/src/tcpx_fabric.c

```diff
@@ -0,0 +1,319 @@
+	if (ret)
+		return ret;
+
+	tcpx_fabric->run_cm_thread = 0;
+	fd_signal_set(&tcpx_fabric->signal);
+
+	if(tcpx_fabric->conn_mgr_thread &&
+	   pthread_join(tcpx_fabric->conn_mgr_thread, NULL))
+		FI_DBG(&tcpx_prov, FI_LOG_FABRIC,
+		       "cm thread failed to join\n");
+	while (!dlist_empty(&tcpx_fabric->fd_list)) {
+		fd_entry = tcpx_fabric->fd_list.next;
+		fd_info_entry = (container_of(fd_entry,
+					struct poll_fd_info,
+					entry));
+		free(fd_info_entry);
+	}
+
+	fastlock_destroy(&tcpx_fabric->fd_list_lock);
+	fd_signal_free(&tcpx_fabric->signal);
+
+	free(tcpx_fabric);
+	return 0;
+}
+
+struct fi_ops tcpx_fabric_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = tcpx_fabric_close,
+	.bind = fi_no_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static void fd_list_delete_item(int index,
+				struct pollfd *poll_fds,
+				struct poll_fd_info *fd_info,
+				int *nfds)
+{
+	poll_fds[index] = poll_fds[(*nfds)-1];
+	fd_info[index] = fd_info[(*nfds)-1];
+	*nfds = (*nfds) - 1;
+}
+
+static void handle_fd_list(struct tcpx_fabric *tcpx_fabric,
+			   struct pollfd *poll_fds,
+			   struct poll_fd_info *fd_info,
+			   int *nfds)
+{
+	int i;
+	struct dlist_entry *fd_entry;
+	struct poll_fd_info *fd_info_entry;
+	struct tcpx_pep *tcpx_pep;
+
+	fastlock_acquire(&tcpx_fabric->fd_list_lock);
+	while (!dlist_empty(&tcpx_fabric->fd_list)) {
+		fd_entry = tcpx_fabric->fd_list.next;
+		fd_info_entry = (container_of(fd_entry,
+					struct poll_fd_info,
+					entry));
+		if (*nfds < 1024) {
+			for (i =0 ; i < *nfds ; i++) {
+				if (poll_fds[i].fd == fd_info_entry->sock_fd) {
+					if ((fd_info_entry->flags & TCPX_SOCK_DEL) &&
+					    (FI_CLASS_PEP == fd_info_entry->fid->fclass)) {
+						fd_list_delete_item(i,poll_fds, fd_info, nfds);
+						tcpx_pep = container_of(fd_info_entry->fid,
+									struct tcpx_pep,
+									util_pep.pep_fid.fid);
+						tcpx_pep->cm.sock_fd_closed = 1;
+					}
+					dlist_remove(&tcpx_fabric->fd_list);
+					free(fd_info_entry);
+					continue;
+				}
+			}
+			fd_info[*nfds] = *fd_info_entry;
+			poll_fds[*nfds].fd = fd_info[*nfds].sock_fd;
+			switch (fd_info[*nfds].fid->fclass) {
+			case FI_CLASS_EP:
+				poll_fds[*nfds].events = POLLOUT;
+				(*nfds)++;
+				break;
+			case FI_CLASS_PEP:
+				poll_fds[*nfds].events = POLLIN;
+				(*nfds)++;
+				break;
+			default:
+				FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+					"invalid fd\n");
+			}
+			dlist_remove(&tcpx_fabric->fd_list);
+			free(fd_info_entry);
+		}
+	}
+	fastlock_release(&tcpx_fabric->fd_list_lock);
+}
+
+static void handle_fd_events(struct tcpx_fabric *fabric,
+			   struct pollfd *poll_fds,
+			   struct poll_fd_info *fd_info,
+			   int *nfds)
+{
+	int iter, status, ret;
+	socklen_t len;
+	struct tcpx_conn_handle *handle;
+	struct util_ep *util_ep;
+	struct tcpx_pep *tcpx_pep;
+	struct fi_eq_cm_entry eq_entry;
+
+	for (iter = 0; iter < *nfds ; iter++) {
+		switch (fd_info[iter].fid->fclass) {
+		case FI_CLASS_EP:
```

Comment:
You can check for revent first and remove the switch statement.  POLLOUT only applies to EPs and POLLIN only to PEPs.  You could assert that in the code.